karaf/osgi-jta/src/main/java/org/jboss/narayana/quickstart/osgi/RecoveryExample.java

```diff
@@ -0,0 +1,177 @@
+import org.jboss.narayana.quickstart.xa.DummyXAResource;
+import org.jboss.narayana.quickstart.xa.DummyXAResourceRecovery;
+import org.jboss.tm.XAResourceRecovery;
+import org.osgi.framework.BundleContext;
+
+import javax.naming.Context;
+import javax.naming.InitialContext;
+import javax.sql.DataSource;
+import javax.transaction.Transaction;
+import javax.transaction.TransactionManager;
+import java.io.File;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+/**
+ * @author <a href="mailto:zfeng@redhat.com">Amos Feng</a>
+ */
+
+@Command(scope = "narayana-quickstart", name = "testRecovery", description="The JTA Recovery Example")
+@Service
+public class RecoveryExample implements Action {
+    @Reference
+    private BundleContext context;
+
+    @Reference
+    private TransactionManager tm;
+
+    @Option(name = "-f", aliases = "--force", description = "force the example to crash", required = false, multiValued = false)
+    boolean force = false;
+
+    @Override
+    public Object execute() throws Exception {
+        if (force) {
+            System.out.println("testRecovery generate something to recovery");
+            if (tm == null) {
+                System.out.println("Can not get transaction manager");
+            } else {
+                DummyXAResource xa = new DummyXAResource(DummyXAResource.faultType.HALT);
+                Context ctx = new InitialContext();
+                DataSource ds = (DataSource)ctx.lookup("osgi:service/test");
+                Connection conn = null;
+                Statement stmt = null;
+
+                try {
+                    tm.begin();
+                    Transaction transaction = tm.getTransaction();
+                    transaction.enlistResource(xa);
+
+                    conn = ds.getConnection();
+                    stmt = conn.createStatement();
+                    stmt.execute("CREATE TABLE IF NOT EXISTS example ( id varchar(10), name varchar(32) )");
+                    stmt.execute("insert into example values (1, 'recovery')");
+
+                    tm.commit();
+                } catch (SQLException e) {
+                    System.out.println(e);
+                    tm.rollback();
+                } finally {
+                    stmt.close();
+                    conn.close();
+                }
+            }
+        } else {
+            File file = new File("DummyXAResource/");
+            if (file.exists() && file.isDirectory() && file.listFiles().length > 0) {
+                System.out.println("register the DummyXAResourceRecovery");
+                context.registerService(XAResourceRecovery.class, new DummyXAResourceRecovery(), null);
+
+                int currentCommits = DummyXAResource.getCommitRequests();
+                System.out.println("current commitRequests of the DummyXAResource is " + currentCommits);
+
+                System.out.println("query the database before recovery");
```

Comment:
OK, I think I get your idea. 

It could be better to run the queryDatabse and getCommitRequests() before registering the DummyXAResourceRecovery
The recovery manager could be scanning the in-doubt record and can not find the helper. you could see in the log
```
2016-04-08 10:01:26,528 | WARN  | eriodic Recovery | jta                              |  -  -  | ARJUNA016037: Could not find new XAResource to use for recovering non-serializable XAResource XAResourceRecord < resource:null, txid:< formatId=131077, gtrid_length=29, bqual_length=36, tx_uid=0:ffff7f000001:a4fe:5707105c:59, node_name=1, branch_uid=0:ffff7f000001:a4fe:5707105c:5a, subordinatenodename=null, eis_name=0 >, heuristic: TwoPhaseOutcome.FINISH_OK com.arjuna.ats.internal.jta.resources.arjunacore.XAResourceRecord@10e7f6a6 >
2016-04-08 10:01:26,530 | WARN  | eriodic Recovery | jta                              |  -  -  | ARJUNA016038: No XAResource to recover < formatId=131077, gtrid_length=29, bqual_length=36, tx_uid=0:ffff7f000001:a4fe:5707105c:59, node_name=1, branch_uid=0:ffff7f000001:a4fe:5707105c:5a, subordinatenodename=null, eis_name=0 >
```

and I think we can be sure that the recovery has not happened. After registering the DummyXAResourceRecovery and wait till the recovery happens.